From: David Sterba <dsterba@suse.cz>
Date: Fri, 10 Feb 2012 18:54:01 +0100
Patch-mainline: pending
References: FATE#306586 bnc#746373
Subject: [PATCH] btrfs: fix permissions of new subvolume

The mode is 0700, which prevents non-root users to access the volume. This
causes trouble when root filesystem is btrfs living in a separate subvolume,
then eg. a daemon or process with dropped privileges cannot access files.

Set the mode to S_IRWXUGO and take umask into account, similar to what mkdir
does. Snapshots are not affected by this change, they'll inherit mode from the
original subvolume.

Signed-off-by: David Sterba <dsterba@suse.cz>
---
 fs/btrfs/inode.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/fs/btrfs/inode.c
+++ b/fs/btrfs/inode.c
@@ -6865,9 +6865,10 @@ int btrfs_create_subvol_root(struct btrf
 	struct inode *inode;
 	int err;
 	u64 index = 0;
+	int mode = S_IRWXUGO & ~current_umask();
 
 	inode = btrfs_new_inode(trans, new_root, NULL, "..", 2, new_dirid,
-				new_dirid, S_IFDIR | 0700, &index);
+				new_dirid, S_IFDIR | mode, &index);
 	if (IS_ERR(inode))
 		return PTR_ERR(inode);
 	inode->i_op = &btrfs_dir_inode_operations;
